jvector-base/src/main/java/io/github/jbellis/jvector/graph/ScoreTracker.java

```diff
@@ -16,13 +16,54 @@
 
 package io.github.jbellis.jvector.graph;
 
+import io.github.jbellis.jvector.util.ArrayUtil;
 import io.github.jbellis.jvector.util.BoundedLongHeap;
 import org.apache.commons.math3.stat.StatUtils;
 
 import static io.github.jbellis.jvector.util.NumericUtils.floatToSortableInt;
 import static io.github.jbellis.jvector.util.NumericUtils.sortableIntToFloat;
 
 interface ScoreTracker {
+    class ScoreTrackerFactory {
+        private TwoPhaseTracker twoPhaseTracker;
+        private RelaxedMonotonicityTracker relaxedMonotonicityTracker;
+        private NoOpTracker noOpTracker;
+
+        ScoreTrackerFactory() {
+            twoPhaseTracker = null;
+            relaxedMonotonicityTracker = null;
+            noOpTracker = null;
+        }
+
+        public ScoreTracker getScoreTracker(boolean pruneSearch, int rerankK, float threshold) {
+            // track scores to predict when we are done with threshold queries
+            ScoreTracker scoreTracker;
```

Comment:
nit: use final to help compiler make sure it's assigned at the end of the if conditions.